File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if root == None:
            return []
        
        stack = deque()
        ans = []
        
        curr = root
        
        while curr != None or stack:
            while curr != None:
                stack.append(curr)
                curr = curr.left
            
            currNode = stack.pop()
            ans.append(currNode.val)
            
            curr = currNode.right
            
        return ans
        
        
        

File: 94-binary-tree-inorder-traversal/94-binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        return solve(root);
    }
private:
    vector<int> solve(TreeNode* root) {
        stack<TreeNode*> stack;
        vector<int> ans;
        TreeNode* tmp = root;
        
        while(tmp != nullptr || !stack.empty()) {
            while(tmp != nullptr) {
                stack.push(tmp);
                tmp = tmp -> left;
            }
            tmp = stack.top();
            stack.pop();
            ans.push_back(tmp -> val);
            
            tmp = tmp -> right;
        }
        return ans;
    }
};